cmd/kubeapps-apis/plugins/kapp_controller/packages/v1alpha1/server_utils_test.go

```diff
@@ -0,0 +1,524 @@
+		{"Reconciling", kappctrlv1alpha1.AppConditionType("Reconciling"), "Reconciling"},
+		{"Unknown", kappctrlv1alpha1.AppConditionType("foo"), "Unknown"},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			userReason := userReasonForKappStatus(tt.status)
+			if want, got := tt.expectedUserReason, userReason; !cmp.Equal(want, got) {
+				t.Errorf("in %s: mismatch (-want +got):\n%s", tt.name, cmp.Diff(want, got))
+			}
+		})
+	}
+}
+
+func TestErrorByStatus(t *testing.T) {
+	tests := []struct {
+		name        string
+		verb        string
+		resource    string
+		identifier  string
+		err         error
+		expectedErr error
+	}{
+		{"error msg for all resources ", "get", "my-resource", "", status.Errorf(codes.InvalidArgument, "boom!"), status.Errorf(codes.Internal, "unable to get the my-resource 'all' due to 'rpc error: code = InvalidArgument desc = boom!'")},
+		{"error msg for a single resources ", "get", "my-resource", "my-id", status.Errorf(codes.InvalidArgument, "boom!"), status.Errorf(codes.Internal, "unable to get the my-resource 'my-id' due to 'rpc error: code = InvalidArgument desc = boom!'")},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			err := errorByStatus(tt.verb, tt.resource, tt.identifier, tt.err)
+			if got, want := err.Error(), tt.expectedErr.Error(); !cmp.Equal(want, got) {
+				t.Errorf("in %s: mismatch (-want +got):\n%s", tt.name, cmp.Diff(want, got))
+			}
+		})
+	}
+}
+
+func TestExtractValue(t *testing.T) {
+	tests := []struct {
+		name     string
+		body     string
+		key      string
+		expected string
+	}{
+		{"retrieves a top level key", `{"a": "foo"}`, "a", " foo"},
+		{"returns '' in nested keys", `{"a": { b: "foo"}}`, "b", ""},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			values := extractValue(tt.body, tt.key)
+			if !cmp.Equal(tt.expected, values) {
+				t.Errorf("mismatch in '%s': %s", tt.name, cmp.Diff(tt.expected, values))
+			}
+		})
+	}
+}
+
+func TestDefaultValuesFromSchema(t *testing.T) {
+	tests := []struct {
+		name     string
+		schema   []byte
+		expected string
+	}{
+		{"schema with and without defaults", []byte(`properties:
+  valueWithDefault:
+    default: 80
+    description: Value with default
+    type: integer
+  missingDefaultInteger:
+    description: Missing default integer
+    type: integer
+  missingDefaultNumber:
+    description: Missing default number
+    type: number
+  missingDefaultString:
+    description: Missing default string
+    type: string
+  missingDefaultBoolean:
+    description: Missing default boolean
+    type: boolean
+  missingDefaultArray:
+    description: Missing default array
+    type: array
+  missingDefaultObject:
+    description: Missing default object
+    type: object`,
+		),
+			`missingDefaultArray: []
+missingDefaultBoolean: false
+missingDefaultInteger: 0
+missingDefaultNumber: 0
+missingDefaultObject: {}
+missingDefaultString: ""
+valueWithDefault: 80
+`},
```

Comment:
Looks like this should be 7 test cases (you've even added the infra for iterating the test cases) that you've put in one test-case? (ie. it will fail if any of the parts fail). Not a big deal and fine as is, just unexpected.

Reply:
haha, you're right; I started writing the different cases at the same time, but I forgot to split them apart. Done (+ few additional test cases more)